metals/src/main/scala/scala/meta/internal/metals/Compilations.scala

```diff
@@ -26,18 +30,29 @@ final class Compilations(
     compileWorksheets: Seq[AbsolutePath] => Future[Unit],
     onStartCompilation: () => Unit,
 )(implicit ec: ExecutionContext) {
-
+  private val compileTimeout: Timeout =
+    Timeout("compile", Duration(10, TimeUnit.MINUTES))
   // we are maintaining a separate queue for cascade compilation since those must happen ASAP
   private val compileBatch =
     new BatchedFunction[
       b.BuildTargetIdentifier,
       Map[BuildTargetIdentifier, b.CompileResult],
-    ](compile, "compileBatch", shouldLogQueue = true, Some(Map.empty))
+    ](
+      compile(timeout = Some(compileTimeout)),
+      "compileBatch",
+      shouldLogQueue = true,
+      Some(Map.empty),
+    )
   private val cascadeBatch =
     new BatchedFunction[
       b.BuildTargetIdentifier,
       Map[BuildTargetIdentifier, b.CompileResult],
-    ](compile, "cascadeBatch", shouldLogQueue = true, Some(Map.empty))
+    ](
+      compile(timeout = None),
+      "cascadeBatch",
```

Comment:
Should we cancel this if we cancel normal compilation also? Since most likely this will also not finish. Cascade compile is normal compile, but on more targets.

Reply:
I set the minimum timeout to 15 min for that.